File: 02주차 구간합/김동주/boj_24838.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <vector>

#define MOD (1000000007)
#define MAX_N (50000)
#define MAX_M (200000)

using namespace std;


long long madd(long long a, long long b, long long mod) {
    return ((a % mod) + (b % mod)) % mod;
}

long long mmul(long long a, long long b, long long mod) {
    return ((a % mod) * (b % mod)) % mod;
}

struct edge {
    int index;
    int diff;
};


bool cmp_edges(const struct edge &p1, const struct edge &p2) {
    return p1.index < p2.index;
}


long long FACTORIAL[MAX_M+1];

int n;
int m;
int A[MAX_N+1];
struct edge edges[2*(MAX_M+1)];
long long frequency[MAX_M+1];


void testcase() {
    cin >> n >> m;

    // O(n)
    for (int i = 0; i < n; i++) {
        cin >> A[i];
    }

    // O(m)
    int n_edges = 0;
    for (int j = 0; j < m; j++) {
        int x, y;

        cin >> x;
        edges[n_edges].index = x;
        edges[n_edges].diff = 1;
        n_edges++;

        cin >> y;
        edges[n_edges].index = y+1;
        edges[n_edges].diff = -1;
        n_edges++;
    }


    // O(n log n + m log m)
    sort(A, A+n);
    sort(edges, edges+n_edges, cmp_edges);

    // O(m)
    memset(frequency, 0, sizeof(frequency));

    // O(m)
    int last_index = 1;
    int last_n_ranges = 0;
    for (int i = 0; i < n_edges; i++) {
        frequency[last_n_ranges] += (edges[i].index - last_index);
        last_n_ranges += edges[i].diff;
        last_index = edges[i].index;
    }
    frequency[0] += (n + 1 - last_index);

    // O(n+m)
    long long max_s = 0;
    long long n_cases = 1;
    int index = n-1;
    for (int n_ranges = m; n_ranges >= 0; n_ranges--) {
        // n_ranges개의 구간에 포함된 원소가 n_elements개.
        int n_elements = frequency[n_ranges];
        for (int i = 0; i < n_elements; i++) {
            max_s += n_ranges * A[index--];
        }
        n_cases = mmul(n_cases, FACTORIAL[n_elements], MOD);
    }

    cout << max_s << ' ' << n_cases << endl;
}

void init() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    // 팩토리얼을 미리 구해놓는다.
    FACTORIAL[0] = 1;
    for (long long i = 1; i <= MAX_M; i++) {
        FACTORIAL[i] = mmul(FACTORIAL[i-1], i, MOD);
    }
}

int main() {
    int T;
    init();
    for (cin >> T; T; T--) testcase();
    return 0;
}

File: 02주차 구간합/김동주/boj_24838.py
from sys import stdin
from sys import stdout


MOD = int(1e9+7)
MAX_N = 50000
MOD_FACTORIAL = []


"""
나머지의 분배법칙 이용

    (A * B) mod C = (A mod C * B mod C) mod C

    https://www.khanacademy.org/computing/computer-science/cryptography/modarithmetic/a/modular-multiplication
"""
def madd(a: int, b: int, mod=MOD) -> int:
    return ((a % mod) + (b % mod)) % mod


def mmul(a: int, b: int, mod=MOD) -> int:
    return ((a % mod) * (b % mod)) % mod


def init_mod_factorial():
    # O(n)
    MOD_FACTORIAL.append(1)
    for i in range(1, MAX_N+1):
        MOD_FACTORIAL.append(mmul(MOD_FACTORIAL[-1], i))


def testcase():
    # 핵심 아이디어: 가장 많이 포함된 구간에 큰 숫자 배정
    # 나머지연산은 덧셈, 곱셈에 대해 분배법칙이 적용됨에 주목할 것.

    n, m = map(int, stdin.readline().split())

    # O(n log n)
    A = list(map(int, stdin.readline().split()))
    A.sort()

    # O(m)
    events = []
    for j in range(m):
        x, y = map(int, stdin.readline().split())
        events.append((x, +1)) # rising edge
        events.append((y+1, -1)) # falling edge
    # O(m log m)
    events.sort()

    # O(m)
    frequency = [0] * (m + 1)
    last_index = 1
    last_n_ranges = 0
    for index, diff in events:
        frequency[last_n_ranges] += (index - last_index)
        last_n_ranges += diff
        last_index = index
    frequency[0] += (n + 1 - last_index)

    # O(n+m)
    # (A는 오름차순으로 정렬되어 있음.)
    max_s = 0
    n_cases = 1
    for n_ranges, n_elements in reversed(list(enumerate(frequency))):
        # n_ranges개의 구간에 포함된 원소가 n_elements개.
        for i in range(n_elements):
            max_s += n_ranges * A.pop()
        n_cases = mmul(n_cases, MOD_FACTORIAL[n_elements]) # 원소를 배치하는 경우의 수

    stdout.write(f'{max_s} {n_cases}\n')


if __name__ == "__main__":
    init_mod_factorial()
    for t in range(int(stdin.readline())):
        testcase()
